launch_ros/launch_ros/ros_adapters.py

```diff
@@ -0,0 +1,138 @@
+          to pass down to rclpy.init.
+          The context is expected to have already been initialized by the caller
+          using rclpy.init().
+        :param: autostart Whether to start adapter on construction or not.
+        """
+        self.__argv = argv
+        self.__context = context
+        self.__node = None
+        self.__executor = None
+        self.__is_running = False
+
+        if autostart:
+            self.start()
+
+    def start(self):
+        """Start ROS adapter."""
+        if self.__is_running:
+            raise RuntimeError('Cannot start a ROS adapter that is already running')
+        try:
+            if self.__context is not None:
+                rclpy.init(args=self.__argv, context=self.__context)
+        except RuntimeError as exc:
+            if 'rcl_init called while already initialized' in str(exc):
+                pass
+            raise
+        self.__node = rclpy.create_node(
+            'launch_ros_{}'.format(os.getpid()),
+            context=self.__context
+        )
+        self.__is_running = True
+        self.__executor = SingleThreadedExecutor(context=self.__context)
+        self.__executor_thread = threading.Thread(target=self._run)
+        self.__executor_thread.start()
+
+    def _run(self):
+        try:
+            self.__executor.add_node(self.__node)
+            while self.__is_running:
+                # TODO(wjwwood): switch this to `spin()` when it considers
+                #   asynchronously added subscriptions.
+                #   see: https://github.com/ros2/rclpy/issues/188
+                self.__executor.spin_once(timeout_sec=1.0)
+        except KeyboardInterrupt:
+            pass
+        finally:
+            self.__executor.remove_node(self.__node)
+
+    def shutdown(self):
+        """Shutdown ROS adapter."""
+        if not self.__is_running:
+            raise RuntimeError('Cannot shutdown a ROS adapter that is not running')
+        self.__executor_thread.join()
+        self.__node.destroy_node()
+        rclpy.shutdown(context=self.__context)
+        self.__is_running = False
+
+    @property
+    def argv(self):
+        return self.__argv
+
+    @property
+    def context(self):
+        return self.__context
+
+    @property
+    def node(self):
+        return self.__node
+
+    @property
+    def executor(self):
+        return self.__executor
+
+
+def get_ros_adapter(*, context: launch.LaunchContext):
+    """
+    Get the ROS adapter managed by the given launch context.
+
+    If no adapter is found, one will be created.
+    """
+    if not hasattr(context.locals, 'ros_adapter'):
+        ros_adapter = ROSAdapter()
+        context.extend_globals({'ros_adapter': ros_adapter})
+        context.register_event_handler(launch.event_handlers.OnShutdown(
+            on_shutdown=lambda *args, **kwargs: ros_adapter.shutdown()
+        ))
+    return context.locals['ros_adapter']
```

Comment:
Should we consider making this thread-safe? I'm just imaging some action creating a thread and trying to access this from the thread and another action accessing it from the asyncio run loop in a different thread, at the same time.

Reply:
Hmm, I had the impression `launch` wasn't thread-safe i.e. that using the same launch context across `asyncio` loops in separate threads wasn't supported. We can lock it otherwise.

Reply:
Parts of it are, if we think a function may be called outside of the asyncio loop (like from a user thread) then it should be thread-safe. I'm not talking about calling the run loop for asyncio in multiple threads concurrently, that is definitely not supported.

Reply:
I see. The thing is that something like:

> I'm just imaging some action creating a thread and trying to access this from the thread and another action accessing it from the asyncio run loop in a different thread, at the same time.

won't work properly because `launch.Context` itself isn't thread-safe e.g. one cannot emit an event or affect context globals or locals in a thread-safe manner. So I'd rather push for that in a follow-up PR i.e. make `launch.Context` thread-safe and then make this function thread-safe. WDYT?

Reply:
The context is not thread-safe, but things that use it are, e.g.: https://github.com/ros2/launch/blob/0d040023306843d71b6316b160fab7b219f41892/launch/launch/launch_service.py#L124-L144

I'd say I agree with you if the user or other actions were accessing the context directly, but they're not, they are calling this function which happens to use the context as an implementation detail.

I think it should still be thread-safe to be honest.

Reply:
Hmm, I think `launch.LaunchService` case is different, as that's API for `launch` client code as opposed to API used within `launch.Action` subclasses, where pretty much everything is not thread-safe. 

But forget all that, how would you make these functions thread-safe without doing the same for `launch.LaunchContext`? I don't think a global lock for these functions is a proper solution, because (a) two threads using two different `launch.LaunchContext` instances may contend with each other for no reason, and (b) `launch.LaunchContext` global dict itself isn't thread-safe (which may be solved with such a lock, yes, but if and only if no other API that could potentially run concurrently attempts to do the same).

Reply:
Why not just a lock in this function? Either a static (global) one or a one owned by this class?

I'm not trying to thread-safe the context, but I do expect other actions using the node or context to access it through this class's methods.

Reply:
> Why not just a lock in this function? Either a static (global) one or a one owned by this class?

Because of the reasons exposed above.

Reply:
Sorry I'll try to address the two points directly.

> (a) two threads using two different launch.LaunchContext instances may contend with each other for no reason

Is that actually a problem? The scope and time bounds on this function are straightforward, so I don't think there's a deadlock problem. Are you concerned about performance due to lock contention?

> (b) launch.LaunchContext global dict itself isn't thread-safe (which may be solved with such a lock, yes, but if and only if no other API that could potentially run concurrently attempts to do the same).

Of course if someone trys to access this fields in the context directly it may be an issue, but I thought the idea was to always access this field of the context (`context.locals.ros_adapter`) via this method right?

Do you mean issues when someone asynchronously calls this function while someone else concurrently accesses a _different_ field in the context?

---

I'm happy to drop this if you don't think it's a concern, my concern is definitely academic for the moment.

Reply:
> Are you concerned about performance due to lock contention?

A bit of that, a bit of OCD :)

> Do you mean issues when someone asynchronously calls this function while someone else concurrently accesses a different field in the context?

Precisely. In particular, `launch.LaunchContext.extend_locals()` and `launch.LaunchContext.extend_globals()` do not seem safe to use concurrently as per this [Python 3.8 FAQ](https://docs.python.org/3.8/faq/library.html#what-kinds-of-global-value-mutation-are-thread-safe) (see disclaimer at the bottom, due to `__del__` being called as a side effect of `dict.update`). I give you that the chance of actually hitting a problem is (likely very) low though.

----

If you think there's no risk of this backfiring in the future, I'll add that lock.

Reply:
We can leave it out. It would be nice to mention it explicitly in the docstring, as I feel this class is more likely to be used concurrently (like in the callback of a subscription or something) than other context functions.

Reply:
Done in 63ff30f88337fe90acbb114063bd9ed835203b7f.